nifi-nar-bundles/nifi-irc-bundle/nifi-irc-processors/src/main/java/org/apache/nifi/processors/irc/PublishIRC.java

```diff
@@ -0,0 +1,139 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.processors.irc;
+
+import org.apache.nifi.annotation.behavior.InputRequirement;
+import org.apache.nifi.annotation.behavior.TriggerSerially;
+import org.apache.nifi.annotation.behavior.WritesAttribute;
+import org.apache.nifi.annotation.behavior.WritesAttributes;
+import org.apache.nifi.annotation.documentation.CapabilityDescription;
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.annotation.lifecycle.OnStopped;
+import org.apache.nifi.flowfile.FlowFile;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.ProcessSessionFactory;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.io.InputStreamCallback;
+import org.apache.nifi.processors.irc.handlers.PublisherEventHandler;
+import org.apache.nifi.stream.io.StreamUtils;
+import org.apache.nifi.util.StopWatch;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.TimeUnit;
+
+@Tags({"publish", "irc"})
+@TriggerSerially
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+
+@CapabilityDescription("This processor implements a IRC client allowing nifi to publish flowfile contents into a " +
+        "predefined IRC channel. " + "\n" +
+        "IMPORTANT NOTE: Due to the nature of the IRC protocol, no delivery guarantees are offered. USE WITH CARE")
+
+@WritesAttributes({
+        @WritesAttribute(attribute = "irc.sender", description = "The IRC nickname source user of the message"),
+        @WritesAttribute(attribute = "irc.channel", description = "The channel from where the message was received "),
+        @WritesAttribute(attribute = "irc.server", description = "The values IRC channel where the message was received from")})
+public class PublishIRC extends AbstractIRCProcessor {
+
+    private volatile PublisherEventHandler eventHandler;
+
+    @OnStopped
+    public void onUnscheduled(ProcessContext context) {
+        clearSetup(client, eventHandler);
+        client = null;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSessionFactory sessionFactory) throws ProcessException {
+        ProcessSession session = sessionFactory.createSession();
+
+
+
+        if (client == null) {
+            // Initialise the handler that will be provided to the setupClient super method
+            this.eventHandler = new PublisherEventHandler(context, sessionFactory, getLogger());
+            this.client = ircClientService.getClient();
```

Comment:
I've not looked into this particular processor in great detail but doing such a thing in onTrigger instead of onScheduled makes sense as a lazy initialization model.  First, onTrigger happens once there is really data to send so it is nicer in cases where there isn't active flow . Second, it can make reconnection logic more natural as there is always a "do i have a connection, if not make one" sort of logic and it happens in the start of ontrigger once there is a flowfile.